core/src/Kokkos_ReductionIdentity.hpp

```diff
@@ -16,421 +16,71 @@
-    return static_cast<unsigned long long>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long min() {
-    return ULLONG_MAX;
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long bor() {
-    return static_cast<unsigned long long>(0x0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long band() {
-    return ~static_cast<unsigned long long>(0x0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long lor() {
-    return static_cast<unsigned long long>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long land() {
-    return static_cast<unsigned long long>(1);
-  }
-};
-
-template <>
-struct reduction_identity<float> {
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static float sum() {
-    return static_cast<float>(0.0f);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static float prod() {
-    return static_cast<float>(1.0f);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static float max() {
-#if __FINITE_MATH_ONLY__
-    return -FLT_MAX;
-#else
-    return -HUGE_VALF;
-#endif
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static float min() {
-#if __FINITE_MATH_ONLY__
-    return FLT_MAX;
-#else
-    return HUGE_VALF;
-#endif
-  }
-};
-
-template <>
-struct reduction_identity<double> {
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static double sum() {
-    return static_cast<double>(0.0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static double prod() {
-    return static_cast<double>(1.0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static double max() {
+template <typename Floating>
+  requires(std::floating_point<Floating> && sizeof(Floating) <= sizeof(double))
+struct reduction_identity<Floating> {
+ private:
+  static constexpr auto inf =
 #if __FINITE_MATH_ONLY__
-    return -DBL_MAX;
+      std::numeric_limits<Floating>::max();
 #else
-    return -HUGE_VAL;
+      std::numeric_limits<Floating>::infinity();
 #endif
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static double min() {
-#if __FINITE_MATH_ONLY__
-    return DBL_MAX;
-#else
-    return HUGE_VAL;
-#endif
-  }
+ public:
+  KOKKOS_FUNCTION constexpr static Floating sum() noexcept { return 0; }
+  KOKKOS_FUNCTION constexpr static Floating prod() noexcept { return 1; }
+  KOKKOS_FUNCTION constexpr static Floating max() noexcept { return -inf; }
+  KOKKOS_FUNCTION constexpr static Floating min() noexcept { return +inf; }
 };
 
 // No __host__ __device__ annotation because long double treated as double in
 // device code.  May be revisited later if that is not true any more.
-template <>
-struct reduction_identity<long double> {
-  constexpr static long double sum() { return static_cast<long double>(0.0); }
-  constexpr static long double prod() { return static_cast<long double>(1.0); }
-
-  constexpr static long double max() {
+template <typename Floating>
+  requires(std::floating_point<Floating> && sizeof(Floating) > sizeof(double))
```

Comment:
The comment does match the template anymore. Also `__fp128` is usable on the device according to https://docs.nvidia.com/cuda/cuda-math-api/cuda_math_api/group__CUDA__MATH__QUAD.html.
I'd just explicitly specialize for `long double`.

Reply:
This PR is already does enough things.
We can investigate adding support for `__float128` on the device side for GPUs that support it later.

Reply:
The comment still doesn't match the template anymore.

Reply:
I actually happen to think that it conveys the idea just fine